addons/mail/models/mail_message.py

```diff
@@ -848,12 +857,145 @@ def _notify(self, force_send=False, send_after_commit=True, user_signature=True)
+                 * reject       - message will be deleted from the database without broadcast
+                                  an email sent to the author with an explanation that the moderators can edit.
+                 * discard      - message will be deleted from the database without broadcast.
+                 * allow        - add email address to white list people of specific channel,
+                                  so that next time if a message come from same email address on same channel,
+                                  it will be automatically broadcasted to relevant channels without any approval from moderator.
+                 * ban          - add email address to black list of emails for the specific channel.
+                                  From next time, a person sending a message using that email address will not need moderation.
+                                  message_post will not create messages with the corresponding expeditor.
+        """
+        # For now, his method is called on a singleton when decision is 'ban' or 'allow'
+        if decision == 'accept':
+            self._accept_message()
+        elif decision in ['reject', 'discard']:
+            self._notify_deletion_and_unlink()
+        else:
+            channel = self.env['mail.channel'].browse(self.res_id)
+            channel._add_to_moderation_email([self.email_from], decision)
+            self._accept_message(True) if decision == 'allow' else self.notify_deletion_and_unlink(True)
+        return True
+
+    def _accept_message(self, allow=False):
+        messages_to_notify = self
+        if allow:
+            for message in self:
+                messages_to_notify |= self.search([
+                    ('moderation_status', '=', 'pending_moderation'),
+                    ('email_from', '=', message.email_from),
+                    ('model', '=', 'mail.channel'),
+                    ('res_id', '=', message.res_id)
+                    ])
+        messages_to_notify.write({'moderation_status': 'accepted', 'moderator_id': self.env.uid})
+        for message in messages_to_notify:
+            message._notify()
+
+    def _notify_deletion_and_unlink(self, ban=False):
+        """ Notify deletion of messages to their moderators and authors and then delete them.
+        """
+        messages_to_delete = self
+        if ban:
+            for message in self:
+                messages_to_delete |= self.search([
+                    ('moderation_status', '=', 'pending_moderation'),
+                    ('email_from', '=', message.email_from),
+                    ('model', '=', 'mail.channel'),
+                    ('res_id', '=', message.res_id)
+                    ])
+
+        channel_ids = messages_to_delete.mapped('res_id')
+        moderators = messages_to_delete.env['mail.channel'].browse(channel_ids).mapped('moderator_ids')
+        authors = messages_to_delete.mapped('author_id')
+
+        moderators_notifications = [
```

Comment:
I don't really understand the remaining of the method. I would vote for some simplification / cleaning :p .

Reply:
done